design/20220722-publicbundle.md

```diff
@@ -0,0 +1,289 @@
+4. Testing the newly built containers in CI to ensure that the updated base images haven't changed any other
+   functionality. If some containers were running an out-of-support base image, they won't get the updated
+   ca-certificates package, and Charles might need to bump to a new minor version which could be incompatible
+   with his applications.
+5. After rebuilding, he has to roll all deployments of all pods to make sure they're using the new images.
+
+All of the above would have to happen under extreme pressure.
+
+### Risks and Mitigations
+
+#### Bundle Container Updates
+
+This proposal makes it absolutely critical that we have a smooth and rapid release process for the
+"bundle" container. A security incident involving a root certificate would be front-page news and a
+major incident for basically every organisation in the world and we'd have to treat it the same,
+to the point of waking engineers up in the middle of the night to cut a new release if required.
+
+#### Bundle Container Attack Surface
+
+Since this proposal includes a need for some kind of process by which the controller container can
+query the bundle from the bundle container, that necessitates an increase in the attack surface of
+trust as an overall application.
+
+The communication channel between the controller and bundle containers is absolutely security critical;
+if a malicious actor were able to intercept and modify the bundle in-flight it would allow them to insert
+their own root certificate and MitM every connection between any container using the resulting bundle.
+
+This is mitigated by the use of two containers in a Pod - the Kubernetes security model should allow
+for this to be done securely without interference.
+
+## Design Details
+
+### Bundle Container "Server"
+
+The initial version of the bundle container would simply embed the trust bundle into a small Go
+application, similar to what [this PR](https://github.com/cert-manager/trust/pull/36) does.
+
+This Go application would then expose the certificate to the controller pod by writing the bundle,
+to a known location on [a shared volume](https://kubernetes.io/docs/tasks/access-application-cluster/communicate-containers-same-pod-shared-volume/).
+
+While it would likely be sufficient to write a PEM-encoded trust bundle directly onto disk, we can
+future-proof a little by wrapping the bundle into a JSON blob. This way, we can choose to add more
+fields to the JSON blob and be sure to ignore them in the trust controller. As long as we don't
+remove or change the type of existing fields, the same blob should be usable with any client.
+
+The initial blob might look like this:
+
+```json
+{
+	"pemBundle": "<pem-encoded-trust-bundle",
+	"bundleVersion": "some version number",
+	"bundleType": "static|dynamic"
+}
+```
+
+`pemBundle` is the bundle itself in PEM format.
+
+`bundleVersion` is the version of the bundle application which served the bundle, which might be
+useful for logging.
```

Comment:
The version will also be essential for writing to the `status` of the `Bundle` object so that a human can verify that trust has propagated a just-released version of a particular bundle.